### PR TITLE
Conclusiones de entrenamientos de los modelos 

### PR DESCRIPTION
**Pull request by marie-adi**
En este apartado el equipo va a documentar sus modelos y los resultados.

### CODE REVIEW

**Comment by marie-adi** [comment]
📊 **Objetivo del Modelo**

Analizar la carga de enfermedad por trastornos depresivos en distintos países usando un modelo de clasificación K-Nearest Neighbors (KNN), categorizando los países en niveles de depresión: Bajo, Medio y Alto.

🧹 **Preprocesamiento de Datos**
   - Dataset: df_final.csv, que contiene indicadores de salud mental por país y año.
   - Filtro temporal: Se selecciona solo el año más reciente disponible (df['year'].max()).
   - Variable objetivo (y): Clasificación del país según nivel de transtornos depresivos (category).
   - Variables predictoras (X): Todas las columnas relevantes excluyendo year, entity, y otras relacionadas a trastornos no depresivos.
   - Normalización: Estandarización de variables con StandardScaler.

🧠 **Creación de Categorías**
 Se utiliza el percentil de DALYs para dividir los países en tres clases:
- Bajo (primer tercil)
- Medio (segundo tercil)
- Alto (tercer tercil)

imagen

📈 **Visualización de Datos**
 - Top 10 países con mayor nivel de depresión
 - Top 10 países con menor nivel de depresión
 - Mapa de calor de correlación entre variables predictoras
 
top 10 paises mayor depresion
top 10 paises menor depresion
Mapa de calor

🤖 **Entrenamiento del Modelo KNN**
- División del dataset: 80% entrenamiento / 20% test, con stratify para balancear las clases.
- Entrenamiento inicial con k=5 vecinos.
- Métricas evaluadas:
   - Accuracy
   - Matriz de confusión
   - Classification report

🔍 **Optimización del Modelo**
- Búsqueda del mejor k de 1 a 30 mediante visualización de error.
- Entrenamiento final con el k óptimo encontrado.
- Evaluación del overfitting mediante diferencia entre precisión en entrenamiento y prueba.
- Validación cruzada con cross_val_score(cv=5).

tasa de error vs valor k

📚 **Curvas de Aprendizaje**
- Evaluación de underfitting/overfitting según brecha entre entrenamiento y validación cruzada.

imagen
Captura de pantalla 2025-05-10 211121

🧪 **Evaluación Multiclase Avanzada**
- Curvas ROC y cálculo de AUC por clase usando OneVsRestClassifier.
- Curvas Precision-Recall y cálculo de Average Precision Score por clase.

Figure_8
Figure_9

✅ **Resumen de Evaluación**

Captura de pantalla 2025-05-10 183441

**Comment by Jorgeluuu** [comment]
# Modelo de Clasificación SVM para Análisis de Depresión por País

## Descripción del Modelo

Este proyecto implementa un modelo de Máquinas de Soporte Vectorial (SVM) para clasificar países según sus niveles de depresión. El modelo analiza diversos indicadores de salud mental y clasifica a los países en dos categorías:

- **Bajo riesgo**: No presenta niveles significativos de depresión
- **Alto riesgo**: Presenta niveles significativos de depresión

## Implementación

### Preprocesamiento de Datos

1. Se utilizan datos del año 2019 como base para el análisis
2. Se eliminan variables no relevantes como:
   - year
   - anxiety_disorders
   - eating_disorders
   - dalys_eating_disorders
   - dalys_anxiety_disorders
   - entity
3. Los datos son escalados utilizando StandardScaler para normalizar las características

### Configuración del Modelo

El modelo SVM se configuró con los siguientes parámetros de búsqueda:

- **Kernel**: RBF y Linear
- **C**: [0.1, 1, 10, 100]
- **Gamma**: ['scale', 'auto', 0.1, 0.01]
- **Class Weight**: 'balanced'

Se utilizó GridSearchCV para encontrar la mejor combinación de hiperparámetros mediante validación cruzada con 5 pliegues.

## Evaluación del Modelo

### Métricas de Rendimiento

El modelo se evalúa utilizando las siguientes métricas:

1. Precisión general del modelo
2. Reporte de clasificación detallado que incluye:
   - Precisión por clase
   - Recall por clase
   - F1-score por clase

### Análisis de Overfitting

Se realiza un análisis de overfitting comparando:
- Score en entrenamiento
- Score en prueba
- Diferencia entre ambos scores

Se establecen los siguientes criterios:
- Diferencia > 0.1: Overfitting significativo
- Diferencia > 0.05: Overfitting moderado
- Diferencia ≤ 0.05: Sin overfitting significativo

## Visualizaciones

El modelo incluye dos visualizaciones principales:

1. **Resultados de la Búsqueda en Cuadrícula**:
   - Muestra la relación entre el valor de C y la precisión media en validación cruzada
   - Permite identificar los mejores hiperparámetros visualmente

2. **Curva de Aprendizaje**:
   - Muestra la evolución del rendimiento del modelo
   - Compara las puntuaciones de entrenamiento y validación cruzada
   - Ayuda a identificar problemas de bias o varianza

## Uso del Modelo

El modelo se puede utilizar para:

1. Predecir el nivel de riesgo de depresión para un país específico
2. Analizar tendencias globales en salud mental
3. Identificar países que requieren atención especial en términos de salud mental

### Ejemplo de Uso

```python
# Cargar el modelo
with open('pkl/modelo_svm.pkl', 'rb') as archivo:
    modelo = pickle.load(archivo)

# Preparar datos de un país
datos_pais = df_2019[df_2019['entity'] == pais]
caracteristicas = datos_pais.drop(['year', 'anxiety_disorders', 'eating_disorders', 
                                'dalys_eating_disorders', 'dalys_anxiety_disorders', 
                                'entity'], axis=1).values

# Realizar predicción
prediccion = modelo.predict(caracteristicas)
```

## Almacenamiento de Resultados

Los resultados del modelo se almacenan en:

1. **modelo_svm.pkl**: Archivo que contiene el modelo entrenado
2. **predicciones_binarias.json**: Archivo JSON con las predicciones para todos los países

## Conclusiones

El modelo SVM implementado proporciona una herramienta efectiva para:

1. Clasificar países según su riesgo de depresión
2. Identificar patrones globales en salud mental
3. Apoyar la toma de decisiones en políticas de salud mental

La combinación de precisión en la clasificación y la capacidad de interpretación de resultados hace que este modelo sea valioso para el análisis de la salud mental a nivel global.

Estos son los resultados obtenidos del entrenamiento del modelo:

image

image

Estas son las graficas obtenidas durante el entrenamiento del modelo:

Curva de aprendizaje
image

Búsqueda en cuadricula
image

**Comment by odar1997** [comment]
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report, roc_auc_score

# Crear y entrenar modelo
logreg = LogisticRegression(max_iter=1000)  # Aseguramos convergencia
logreg.fit(X_train, y_train)

# Predicciones
y_pred = logreg.predict(X_test)
y_proba = logreg.predict_proba(X_test)[:, 1]  # Probabilidades para clase 1

# Matriz de confusión y reporte
print("Matriz de Confusión:\n", confusion_matrix(y_test, y_pred))
print("\nReporte de Clasificación:\n", classification_report(y_test, y_pred))

# AUC-ROC
print("AUC-ROC:", roc_auc_score(y_test, y_proba))

Matriz de Confusión:
 [[567  46]
 [ 59 558]]

Reporte de Clasificación:
               precision    recall  f1-score   support

           0       0.91      0.92      0.92       613
           1       0.92      0.90      0.91       617

    accuracy                           0.91      1230
   macro avg       0.91      0.91      0.91      1230
weighted avg       0.91      0.91      0.91      1230

AUC-ROC: 0.9688436125968679



Resumen del Notebook

Este notebook se enfoca en realizar un análisis de regresión logística para predecir la prevalencia de trastornos mentales. El objetivo es determinar si la prevalencia de ciertos trastornos (esquizofrenia, depresión, ansiedad, bipolaridad y trastornos alimentarios) puede predecir si un país tiene una alta prevalencia general de trastornos mentales.

Pasos Clave:

Carga y Preprocesamiento de Datos:

Se carga un dataset preprocesado (df_final_encoded.csv).
Se muestra un vistazo a los datos cargados.
Definición de la Variable Objetivo:

Se crea una nueva columna binaria (high_prevalence_country) que indica si la prevalencia de trastornos mentales en un país está por encima de la mediana.
Preparación de Datos para el Modelo:

Se definen las variables predictoras (prevalencia de los trastornos mencionados) y la variable objetivo (high_prevalence_country).
Se dividen los datos en conjuntos de entrenamiento y prueba.
Entrenamiento del Modelo de Regresión Logística:

Se entrena un modelo de regresión logística con los datos de entrenamiento.
Evaluación del Modelo:

Se realizan predicciones con el conjunto de prueba.
Se calcula la matriz de confusión, el informe de clasificación y la curva ROC para evaluar el rendimiento del modelo.
Se visualiza la matriz de confusión.
Análisis a Nivel de País:

Se agregan los datos por país, calculando la prevalencia promedio de los trastornos.
Se entrena un nuevo modelo de regresión logística usando estos datos agregados para predecir la prevalencia a nivel de país.
Se evalúa este modelo de la misma manera que el anterior.
En resumen: El notebook construye y evalúa modelos de regresión logística para predecir la alta prevalencia de trastornos mentales, tanto a nivel de registros individuales como a nivel de país, utilizando la prevalencia de otros trastornos como predictores.

**Comment by abbyenredes** [comment]
## Random forest
image

## curva de aprendizaje
image

## Matriz de *confusión*
image

## Curva ROC
image

##  Mapa de riesgo por país (gráfico de barras)(top 20
image

## Conteo de países por categoría de riesgo
image
